Runtime tabulated v2

celdred opened the pull request:
This adds a Themis interface for TSFC that supports both assembly and interpolation kernels with runtime tabulation. The relevant changes are stubbed out in the Firedrake and UFC interfaces.

wence- commented on tsfc/kernel_interface/themis.py:
```diff
@@ -0,0 +1,411 @@
+        for i, coefficient in enumerate(coefficients):
+            self.coefficient_args.append(
+                self._coefficient(coefficient, "w_%d" % i))
+        self.kernel.coefficient_numbers = tuple(coefficient_numbers)
```
Again, lots of this code is identical to the firedrake kernel builder.

celdred replied:
There is no longer a Themis interface, instead the Firedrake interface (with tabulations added) works for us.

wence- commented on tsfc/kernel_interface/themis.py:
```diff
@@ -0,0 +1,411 @@
+
+        for name_, shape in self.tabulations:
+            args.append(coffee.Decl(self.scalar_type, coffee.Symbol(
+                name_, rank=shape), qualifiers=["const"]))
```
This bit is different.

wence- commented on tsfc/kernel_interface/themis.py:
```diff
@@ -0,0 +1,411 @@
+        # Constant
+
+        # Here we use arrays rather than pointers to fit with how Themis calls kernels
+        funarg = coffee.Decl(scalar_type, coffee.Symbol(name, rank=(1,)), qualifiers=["const"])
```
Can you show the difference in the code you want from what you get? Because I'm not sure we will have this control of the argument list with loopy.

This makes:
```c
const double foo[1];
```
Rather than the current 
```c
const * restrict foo
```
What is wrong with the latter for you?

celdred replied:
I have refactored Themis to use the same kernel arguments as Firedrake (plus tabulations), so the Themis interface is no longer needed.

wence- commented on tsfc/kernel_interface/themis.py:
```diff
@@ -0,0 +1,411 @@
+    if not interior_facet:
+        funarg = coffee.Decl(scalar_type, coffee.Symbol(name, rank=(size,)), qualifiers=["const"])
+    else:
+        funarg = coffee.Decl(scalar_type, coffee.Symbol(name, rank=(2*size,)), qualifiers=["const"])
```
Similarly here.

wence- commented on tsfc/kernel_interface/themis.py:
```diff
@@ -0,0 +1,411 @@
+    funarg = coffee.Decl(scalar_type, coffee.Symbol("A", rank=c_shape))
+    varexp = gem.Variable("A", c_shape)
+    expressions = [expression(gem.view(varexp, *slices)) for slices in slicez]
+    return funarg, prune(expressions)
```
This is identical to the firedrake.py version.

celdred commented on tsfc/driver.py:
```diff
@@ -286,7 +290,10 @@ def compile_expression_at_points(expression, points, coordinates, parameters=Non
+    if interface is None:
+        builder = firedrake_interface.ExpressionKernelBuilder(parameters["scalar_type"])
+    else:
+        builder = interface.ExpressionKernelBuilder(parameters["scalar_type"])
```
Fixed this in the new commit.

celdred commented on tsfc/kernel_interface/firedrake.py:
```diff
@@ -273,14 +273,15 @@ def require_cell_sizes(self):
 
         :arg name: function name
         :arg body: function body (:class:`coffee.Block` node)
+        :arg quadrature rule: quadrature rule (not used, stubbed out for Themis integration)
```
Fixed this in the new commit.

celdred commented on tsfc/kernel_interface/ufc.py:
```diff
@@ -102,14 +102,15 @@ def set_coefficients(self, integral_data, form_data):
 
         :arg name: function name
         :arg body: function body (:class:`coffee.Block` node)
+        :arg quadrature rule: quadrature rule (not used, stubbed out for Themis integration)
```
Fixed this in the new commit.

celdred commented:
I have removed the Themis interface and instead added tabulation registration to the Firedrake interface (which does nothing unless there are runtime tabulated elements in the form, so it shouldn't break anything on the Firedrake end). This gets rid of the code duplication. I also changed the compile_at_points interface argument to take an ExpressionKernelBuilder module (as in compile_integral), rather than an interface file.

celdred commented on tsfc/kernel_interface/themis.py:
```diff
@@ -0,0 +1,411 @@
+        for name_, shape in self.tabulations:
+            args.append(coffee.Decl(self.scalar_type, coffee.Symbol(
+                name_, rank=shape), qualifiers=["const"]))
+# POSSIBLY REGISTER A QUAD RULE HERE?
```
It does not! But the Themis interface is gone :)

celdred commented:
Note: The new compile_at_points requires https://github.com/firedrakeproject/firedrake/commit/db641d6168b1d11699e9e2e1b8e41a89b9ff6a30 in Firedrake (ie ignore the new list of tabulations that is returned). Thoughts?

celdred commented on tsfc/kernel_interface/firedrake.py:
```diff
@@ -265,6 +278,14 @@ def set_coefficients(self, integral_data, form_data):
+            if isinstance(node, gem.Variable) and node.name.startswith("rt_"):
+                tabulations[node.name] = node.shape
+        self.tabulations = tuple(sorted(tabulations.items()))
+        self.kernel.tabulations = tuple(sorted(tabulations.items()))
```
Fixed!

celdred commented on tsfc/kernel_interface/ufc.py:
```diff
@@ -102,16 +102,20 @@ def set_coefficients(self, integral_data, form_data):
         :returns: a COFFEE function definition object
         """
+
+        assert(quadrature_rule is None)
```
Done!

celdred commented:
I tried to squash the commits together, but clearly did not do the correct thing. How exactly do I accomplish this?

miklos1 commented on tsfc/driver.py:
```diff
@@ -252,10 +255,10 @@ def name_multiindex(multiindex, name):
                            parameters["scalar_type"], expressions, split_argument_indices)
 
-    return builder.construct_kernel(kernel_name, body)
+    return builder.construct_kernel(kernel_name, body, quad_rule)
```
This is kind of bugged: `quad_rule` comes from a loop over quadratures, but here outside the loop it only references the value in the last iteration. It works as long as there is only one quadrature in an integral.

celdred replied:
This is very true! It We only need quad_rule in order to figure out which points to tabulate at for runtime tabulation. A better way to do this would be to somehow attach this information to the tabulations themselves. I am not sure how to do this though...

wence- replied:
So what is the kernel interface right now? You make a quad rule so that TSFC knows what size the tabulations are (and that's it?). What part of themis uses the quad rule that comes back in `construct_kernel`?

celdred replied:
I think TSFC already knows the size of the tabulations (they are just GEM nodes?). Themis uses the quad_rule to determine which points to tabulate at (using the shift axis and restriction), since the tabulations don't actually know this....

wence- replied:
Well, what happens is that the quad rule comes in (with some points). This point set is provided to the element for tabulation, which constructs the shape of the tabulation matrix like so:

```python
        shape = tuple(index.extent for index in ps.indices) + self.index_shape + self.value_shape
```

where `ps.indices` are the indices used to index the quadrature rule's point set.

celdred replied:
Is there a way to associate these points with the tabulation (specifically the GEM node) within TSFC/FInAT?

wence- replied:
Right now, FInAT tabulates like this:

```python
        for derivative in range(order + 1):
            for alpha in mis(dimension, derivative):
                name = str.format("rt_{}_{}_{}_{}_{}_{}",
                                  self.variant,
                                  self.degree,
                                  ''.join(map(str, alpha)),
                                  self.shift_axes,
                                  'c' if self.continuous else 'd',
                                  {None: "",
                                   '+': "p",
                                   '-': "m"}[self.restriction])
                result[alpha] = gem.partial_indexed(gem.Variable(name, shape), ps.indices)
```

You would, I suppose, need to introduce a new node in gem. Say `gem.Tabulation`? that behaves like a variable, but carries that information. Arguably, this might things cleaner because the encoding of the tabulation information now doesn't live in the string name (I don't know if you use that at all).

miklos1 commented on tsfc/kernel_interface/ufc.py:
```diff
@@ -102,16 +102,20 @@ def set_coefficients(self, integral_data, form_data):
         :returns: a COFFEE function definition object
         """
+
+        assert quadrature_rule is None
```
This assertion just breaks UFC...

miklos1 commented:
I feel like this is the occasion to make use of dynamical scoping
(see `gem.utils.DynamicallyScoped`).  We could declare a dynamically
scoped variable next to RuntimeTabulated, the FInAT element, and the
FInAT element could record the quadrature points as a mapping from the
element names.

The dynamic scoping would ensure that these quadrature points not stick
around forever in global variables, but live only within a certain
dynamic context.

From Themis, one should then call the form compiler from within a
dynamic-let block.  Once the form compiler returned, but still within
the dynamic-let block, one could inspect and retrieve the quadrature
points that were given to RuntimeTabulated.  Leaving the block, then
everything is automatically cleaned up.

celdred <notifications@github.com> writes:
> celdred commented on this pull request.
>
>
>
>> @@ -252,10 +255,10 @@ def name_multiindex(multiindex, name):
>      body = generate_coffee(impero_c, index_names, parameters["precision"],
>                             parameters["scalar_type"], expressions, split_argument_indices)
>  
> -    return builder.construct_kernel(kernel_name, body)
> +    return builder.construct_kernel(kernel_name, body, quad_rule)
>
> Is there a way to associate these points with the tabulation (specifically the GEM node) within TSFC/FInAT?
>
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/firedrakeproject/tsfc/pull/187#discussion_r269175327

celdred commented:
I prefer the idea of creating a new GEM node (gem.tabulation), that can carry around the information currently in the string AND the actual points that are tabulated at. I will have a go at coding that up once the loopy changes land. 

What would be the advantages of using dynamic scoping instead?

miklos1 commented:
When there are multiple quadratures in the same integral, it is likely
that the same element configuration needs to be tabulated at different
sets of quadrature points, which means that each of those tabulations
need a unique name for the kernel argument.  The dynamic context makes
it easy to inspect what other tabulations have been requested on the
same element configuration, so the naming issue could be solved without
needing to invent an elaborate name encoding of quadrature rules.

celdred <notifications@github.com> writes:
> I prefer the idea of creating a new GEM node (gem.tabulation), that
> can carry around the information currently in the string AND the
> actual points that are tabulated at. I will have a go at coding that
> up once the loopy changes land.
>
> What would be the advantages of using dynamic scoping instead?
>
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/firedrakeproject/tsfc/pull/187#issuecomment-476755617

celdred commented:
Yes this issue has recently been discovered, in the context of CellVolume for facet integrals. Right now each gem Variable node carries information about name (which has variant, derivorder, shiftaxis, degree, continuity and restriction currently in it) and shape, which as you noted is not sufficient to uniquely identify tabulations.

I was planning on creating a new gem node Tabulation, that is a subclass of variable, and holds shape, variant, derivorder, shiftaxis, degree, continuity, restriction and point set in it (with all of these in __front__ except point set). Is this sufficient to unique identify tabulations?

miklos1 commented:
This is, of course, enough to uniquely identify tabulations, but being
`gem.Variable`s (i.e. kernel arguments), they still need to have a
_unique_ name in the generated C code.

Currently, the name includes everything except for the quadrature rule,
which means that different tabulations would have the same name when
multiple quadratures are used.  Clearly, something related to the
quadrature point set has to be added to be able to distinguish between
different tabulations of the same element.

With the dynamic context, one could just number the different
tabulations of the same element.  Without that context, you would need
to come up with a scheme to encode the quadrature rule in the name.

celdred <notifications@github.com> writes:
> Yes this issue has recently been discovered, in the context of
> CellVolume for facet integrals. Right now each gem Variable node
> carries information about name (which has variant, derivorder,
> shiftaxis, degree, continuity and restriction currently in it) and
> shape, which as you noted is not sufficient to uniquely identify
> tabulations.
>
> I was planning on creating a new gem node Tabulation, that is a
> subclass of variable, and holds shape, variant, derivorder, shiftaxis,
> degree, continuity, restriction and point set in it (with all of these
> in __front__ except point set). Is this sufficient to unique identify
> tabulations?
>
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/firedrakeproject/tsfc/pull/187#issuecomment-477609814

celdred commented:
Ah that makes sense, I'll take a look at gem.utils.DynamicallyScoped.

celdred commented:
Uggh it is even uglier than we thought- there are of course multiple copies of the same (identical variant, shiftaxis, degree, continuity and restriction) runtime tabulated elements that are being tabulated (sometimes at the same points and derivorder, sometimes different). Ideally each unique combination of variant, shiftaxis, degree, continuity and restriction plus derivorder/pts should generate the "same" gem Variable/kernel argument.

I looked at gem.utils.DynamicallyScoped, I don't understand what it is doing! I am concerned about maintainability if we go that approach...

What about this idea:
Runtime Tabulated is a singleton (one per set of variant, shiftaxis, degree, continuity and restriction; per invocation of compile_form/compile_at_points), and one dictionary and one counter for each deriv order (0, 1, 2). Every time a tabulation is requested it checks the dictionary for the pts, and if not found, increments the counter and stores the pts. Then it generates a Tabulation node (that is a subclass of Variable) that stores variant, derivorder, shiftaxis, degree, continuity, restriction and pts index, plus the pts themselves.

Thus each combination of variant, shiftaxis, degree, continuity, restriction, derivorder and pts generates the "same" gem Variable/kernel argument.

miklos1 commented:
Dynamical scoping is, in its original sense, a way of resolving variable
references.  By now, lexical scoping has been established as the
standard approach, and for good, since lexical scoping is much more
sensible in the general case.

Nevertheless, there are still some particular use cases, when dynamical
scoping is handy.  However, it is little known, for most popular
languages nowadays lack support for dynamical scoping.  The most popular
ones, that I know of, and have this feature, are Perl and Common Lisp.

Python does not have this feature either, but
gem.utils.DynamicallyScoped provides a library implementation.  Sadly,
that class is not really documented yet.

variable = DynamicallyScoped()

defines a dynamically scoped variable with no global value.

variable = DynamicallyScoped(42)

defines a dynamically scoped variable with 42 as its "default" global
value.  The value of the `variable` should always be referenced as
`variable.value`.  Dynamic extents can be defined using the `.let(...)`
method with 'with' blocks.

An example can enlighten this much better:

```python
var = DynamicallyScoped(13)

def f():
    print(var.value)

f()  # prints 13

with var.let(42):
     # within this block, the value of var is 42

     f()  # prints 42

# leaving the dynamic let block, var reverts to its previous value

f()  # prints 13
```

The same example without default value:


```python
var = DynamicallyScoped()

def f():
    print(var.value)

f()  # error, var unset

with var.let(42):
     f()  # ok, prints 42

f()  # error, var unset
```

About the idea you described.  1) Not quite sure what you mean by
singleton.  2) "per invocation of compile_form/compile_at_points": how
do you ensure that the naming / numbering is _per invocation_?  This is
exactly where dynamic variables might help, because if you wrap your
call to compile_form with a dynamic-let block (`with var.let(value):`),
then you numbering is automatically reset when you leave the block.
3) The rest I agree with.

celdred <notifications@github.com> writes:
> Uggh it is even uglier than we thought- there are of course multiple
> copies of the same (identical variant, shiftaxis, degree, continuity
> and restriction) runtime tabulated elements that are being tabulated
> (sometimes at the same points and derivorder, sometimes
> different). Ideally each unique combination of variant, shiftaxis,
> degree, continuity and restriction plus derivorder/pts should generate
> the "same" gem Variable/kernel argument.
>
> I looked at gem.utils.DynamicallyScoped, I don't understand what it is
> doing! I am concerned about maintainability if we go that approach...
>
> What about this idea:
> Runtime Tabulated is a singleton (one per set of variant, shiftaxis,
> degree, continuity and restriction; per invocation of
> compile_form/compile_at_points), and one dictionary and one counter
> for each deriv order (0, 1, 2). Every time a tabulation is requested
> it checks the dictionary for the pts, and if not found, increments the
> counter and stores the pts. Then it generates a Tabulation node (that
> is a subclass of Variable) that stores variant, derivorder, shiftaxis,
> degree, continuity, restriction and pts index, plus the pts
> themselves.
>
> Thus each combination of variant, shiftaxis, degree, continuity,
> restriction, derivorder and pts generates the "same" gem
> Variable/kernel argument.
>
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/firedrakeproject/tsfc/pull/187#issuecomment-477954369
